### PR TITLE
Add docs for ExternalDartReference

### PR DESCRIPTION
**Pull request by srujzs**
This type was added as part of https://github.com/dart-lang/sdk/issues/55187 to dart:js_interop for a faster alternative to JSBoxedDartObject, so we should add some documentation on it.

### CODE REVIEW

**Review by sigmundch — APPROVED** [review]
Only some suggestions, but LGTM even as is

File: src/content/interop/js-interop/js-types.md [hunk 1]
```diff
@@ -73,7 +82,7 @@ Generally, the conversion table looks like the following:
 <div class="table-wrapper" markdown="1">
 
 | JS type                             | Dart type                                |
```
[the remark this answers is not in the document]
Comment (MaryaBelanger) [reply src/content/interop/js-interop/js-types.md:84]: > I believe we use JS-type and interop-type interchangingly.

We actually don't use them interchangeably. I think JS types are interop types, but not all interop types are JS types. From the [Usage](https://dart.dev/interop/js-interop/usage#interop-types) page:

> Interop types are either a ["JS type"](https://dart-dev--pr5668-addexternaldartreference-45iczn14.web.app/interop/js-interop/js-types) provided by Dart or an [extension type](https://dart-dev--pr5668-addexternaldartreference-45iczn14.web.app/language/extension-types) wrapping an interop type.

File: src/content/interop/js-interop/js-types.md [hunk 2]
```diff
@@ -38,6 +38,15 @@ JS types form a natural type hierarchy:
 
 You can find the definition of each type in the [`dart:js_interop` API docs].
```
Comment (MaryaBelanger) [inline src/content/interop/js-interop/js-types.md:40]: I think the new content should either be 1) primarily in the api docs, or 2) in a breakout section on this page.

It's a sizable amount of digression for the Type hierarchy section. I think either of these options would also allow more room for Siggi's [comment](https://github.com/dart-lang/site-www/pull/5668/files#r1540041570) about elaborating on when to use each.

1) API docs:

Add a sub-bullet under the `JSBoxedDartObject` item in the list (line 38) and put your first two sentences there (slight rewrite for wording):

```markdown
- `JSBoxedDartObject`, which allows users to box and pass Dart values
    opaquely within the same Dart runtime.
  - From Dart 3.4 onwards, the type `ExternalDartReference` in `dart:js_interop` can be used to
    accomplish the same thing, but is *not* a JS type. To learn more about the tradeoffs between
    each option, read [API docs link].
```

And then your second paragraph (lines 44-48) could be a note in the API docs for either or both types (I think both makes sense).

2) Another section on this page:

Add the same sub-bullet in 1), but link to the section on the page instead (`To learn more about the tradeoffs between each option, read [link to section]`).

At the bottom of the page, add a section and include your second paragraph + Siggi's "when a JS type is needed" explanation:

```markdown
## `JSBoxedDartObject` vs `ExternalDartReference`

Both `JSBoxedDartObject` and `ExternalDartReference` pass opaque references to
Dart `Object`s through JavaScript. However, `JSBoxedDartObject` wraps the opaque
reference in a JavaScript object, while `ExternalDartReference` is the reference
itself and therefore is not a JS type. Use `ExternalDartReference` where
performance is needed and `JSBoxedDartObject` where a JS type is needed.

(edit as needed: )
For example, if you just want to pass references opaquely back and forth using the
reference as an identity, then `ExternalDartReference` is a good fit. However, if you want
a reference that can also store data and you want to extend it to add data to that reference,
then we recommend using `JSBoxedDartObject`. 
```

And I'd even add more of the tradeoff details from your issue https://github.com/dart-lang/sdk/issues/55187

---

I don't have a preference, though I'm not too familiar with API doc best practices so I'm not sure if 1) makes sense for the kind of information usually kept there, or if repeating the explanation on both type's pages is bad practice 🤷‍♀️

Comment (srujzs) [reply src/content/interop/js-interop/js-types.md:44]: Thanks for the suggestion! I like the idea of splitting it into a separate section (2) and avoiding polluting the type hierarchy section so I went ahead and did that.

File: src/content/interop/js-interop/js-types.md [hunk 3]
```diff
@@ -252,4 +275,8 @@ TODO: add links (with stable) when ready:
 [`instanceOfString`]: https://api.dart.dev/dev/dart-js_interop/JSAnyUtilityExtension/instanceOfString.html
 [`isA`]: https://api.dart.dev/dev/dart-js_interop/JSAnyUtilityExtension/isA.html
 [#4841]: https://github.com/dart-lang/linter/issues/4841
-[#54025]: https://github.com/dart-lang/sdk/issues/54025
\ No newline at end of file
+[#54025]: https://github.com/dart-lang/sdk/issues/54025
+[`JSBoxedDartObject`]: https://api.dart.dev/main/dart-js_interop/JSBoxedDartObject-extension-type.html
+[`ExternalDartReference`]: https://api.dart.dev/main/dart-js_interop/ExternalDartReference-extension-type.html
+[`toExternalReference`]: https://api.dart.dev/main/dart-js_interop/ObjectToExternalDartReference/toExternalReference.html
+[`toDartObject`]: https://api.dart.dev/main/dart-js_interop/ExternalDartReferenceToObject/toDartObject.html
```
Comment (MaryaBelanger) [inline src/content/interop/js-interop/js-types.md:282]:
```suggestion
[`JSBoxedDartObject`]:  {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-js_interop/JSBoxedDartObject-extension-type.html
[`ExternalDartReference`]:  {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-js_interop/ExternalDartReference-extension-type.html
[`toExternalReference`]:  {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-js_interop/ObjectToExternalDartReference/toExternalReference.html
[`toDartObject`]:  {{site.dart-api}}/{{site.sdkInfo.channel}}/dart-js_interop/ExternalDartReferenceToObject/toDartObject.html
```

The rest of the links were updated in [this PR](https://github.com/dart-lang/site-www/pull/5667/files), so I think you'll get those changes if you just update the branch. This suggestion is to correct the new ones.

**Review by MaryaBelanger — APPROVED** [review]
Thanks for incorporating my suggestions! One more thing about the links but otherwise this looks great

File: src/content/interop/js-interop/js-types.md (hunk 3 above)
Comment (srujzs) [reply src/content/interop/js-interop/js-types.md]: Done, thanks!

**Pull request merged**